File: solutions/csharp/parkinglot/ParkingSpot.cs
using ParkingLot.VehicleTypes;
using System;

namespace ParkingLot
{
    public class ParkingSpot
    {
        private readonly int spotNumber;
        private readonly VehicleType vehicleType;
        private Vehicle parkedVehicle;

        public ParkingSpot(int spotNumber, VehicleType vehicleType)
        {
            this.spotNumber = spotNumber;
            this.vehicleType = vehicleType;
        }

        public bool IsAvailable()
        {
            return parkedVehicle == null;
        }

        public void ParkVehicle(Vehicle vehicle)
        {
            if (IsAvailable() && vehicle.GetVehicleType() == vehicleType)
            {
                parkedVehicle = vehicle;
            }
            else
            {
                throw new ArgumentException("Invalid vehicle type or spot already occupied.");
            }
        }

        public void UnparkVehicle()
        {
            parkedVehicle = null;
        }

        public VehicleType GetVehicleType()
        {
            return vehicleType;
        }

        public Vehicle GetParkedVehicle()
        {
            return parkedVehicle;
        }

        public int GetSpotNumber()
        {
            return spotNumber;
        }
    }
}

File: solutions/golang/parkinglot/parking_spot.go
package parkinglot

type ParkingSpot struct {
	spotNumber    int
	vehicleType   VehicleType
	parkedVehicle Vehicle
}

func NewParkingSpot(spotNumber int, vehicleType VehicleType) *ParkingSpot {
	return &ParkingSpot{spotNumber: spotNumber, vehicleType: vehicleType}
}

func (ps *ParkingSpot) IsAvailable() bool {
	return ps.parkedVehicle == nil
}

func (ps *ParkingSpot) ParkVehicle(vehicle Vehicle) {
	if ps.IsAvailable() && vehicle.GetType() == ps.vehicleType {
		ps.parkedVehicle = vehicle
	}
}

func (ps *ParkingSpot) UnparkVehicle() {
	ps.parkedVehicle = nil
}

func (ps *ParkingSpot) GetSpotNumber() int {
	return ps.spotNumber
}

func (ps *ParkingSpot) GetVehicleType() VehicleType {
	return ps.vehicleType
}

func (ps *ParkingSpot) GetParkedVehicle() Vehicle {
	return ps.parkedVehicle
}
